internal/resources/ephemeral/ephemeral_resources.go

```diff
@@ -0,0 +1,256 @@
+// graph walk, and because it might need periodic "renewing" in order to
+// remain live for the necessary duration.
+type Resources struct {
+	active addrs.Map[addrs.ConfigResource, addrs.Map[addrs.AbsResourceInstance, *resourceInstanceInternal]]
+	mu     sync.Mutex
+
+	// WaitGroup to track renew goroutines
+	wg sync.WaitGroup
+}
+
+func NewResources() *Resources {
+	return &Resources{
+		active: addrs.MakeMap[addrs.ConfigResource, addrs.Map[addrs.AbsResourceInstance, *resourceInstanceInternal]](),
+	}
+}
+
+type ResourceInstanceRegistration struct {
+	Value        cty.Value
+	ConfigBody   hcl.Body
+	Impl         ResourceInstance
+	FirstRenewal *providers.EphemeralRenew
+}
+
+func (r *Resources) RegisterInstance(ctx context.Context, addr addrs.AbsResourceInstance, reg ResourceInstanceRegistration) {
+	if addr.Resource.Resource.Mode != addrs.EphemeralResourceMode {
+		panic(fmt.Sprintf("can't register %s as an ephemeral resource instance", addr))
+	}
+
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	configAddr := addr.ConfigResource()
+	if !r.active.Has(configAddr) {
+		r.active.Put(configAddr, addrs.MakeMap[addrs.AbsResourceInstance, *resourceInstanceInternal]())
+	}
+	ri := &resourceInstanceInternal{
+		value:       reg.Value,
+		configBody:  reg.ConfigBody,
+		impl:        reg.Impl,
+		renewCancel: noopCancel,
+	}
+	if reg.FirstRenewal != nil {
+		ctx, cancel := context.WithCancel(ctx)
+		ri.renewCancel = cancel
+
+		r.wg.Add(1)
+		go ri.handleRenewal(ctx, &r.wg, reg.FirstRenewal)
+	}
+	r.active.Get(configAddr).Put(addr, ri)
+}
+
+func (r *Resources) InstanceValue(addr addrs.AbsResourceInstance) (val cty.Value, live bool) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	configAddr := addr.ConfigResource()
+	insts, ok := r.active.GetOk(configAddr)
+	if !ok {
+		return cty.DynamicVal, false
+	}
+	inst, ok := insts.GetOk(addr)
+	if !ok {
+		return cty.DynamicVal, false
+	}
+	// If renewal has failed then we can't assume that the object is still
+	// live, but we can still return the original value regardless.
+	return inst.value, !inst.renewDiags.HasErrors()
+}
+
+// CloseInstances shuts down any live ephemeral resource instances that are
+// associated with the given resource address.
+//
+// This is the "happy path" way to shut down ephemeral resource instances,
+// intended to be called during the visit to a graph node that depends on
+// all other nodes that might make use of the instances of this ephemeral
+// resource.
+//
+// The runtime should also eventually call [Resources.Close] once the graph
+// walk is complete, to catch any stragglers that we didn't reach for
+// piecemeal shutdown, e.g. due to errors during the graph walk.
+func (r *Resources) CloseInstances(ctx context.Context, configAddr addrs.ConfigResource) tfdiags.Diagnostics {
+	r.mu.Lock()
+
+	// TODO upgrade this to a RWMutex so we don't need to serialize all Close
+	// operations which may require network access
```

Comment:
This looks like a small enough change that may still be worth doing in this PR, or is there some other obvious context I'm missing here?

Reply:
oops, I did mean to do that! not sure how this slipped through as I caught at the other FIXME/TODOs I left 🙄